pkg/webhook/authority/authority.go

```diff
@@ -0,0 +1,395 @@
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package authority
+
+import (
+	"bytes"
+	"crypto"
+	"crypto/rand"
+	"crypto/tls"
+	"crypto/x509"
+	"crypto/x509/pkix"
+	"fmt"
+	"math/big"
+	"sync"
+	"time"
+
+	"github.com/go-logr/logr"
+	corev1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/fields"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/informers"
+	"k8s.io/client-go/kubernetes"
+	coreclientset "k8s.io/client-go/kubernetes/typed/core/v1"
+	corelisters "k8s.io/client-go/listers/core/v1"
+	"k8s.io/client-go/rest"
+	"k8s.io/client-go/tools/cache"
+
+	cmapi "github.com/jetstack/cert-manager/pkg/apis/certmanager/v1alpha2"
+	cmmeta "github.com/jetstack/cert-manager/pkg/internal/apis/meta"
+	"github.com/jetstack/cert-manager/pkg/util/pki"
+)
+
+// DynamicAuthority manages a certificate authority stored in a Secret resource
+// and provides methods to obtain signed leaf certificates.
+// The private key and certificate will be automatically generated, and when
+// nearing expiry, the private key and root certificate will be rotated.
+type DynamicAuthority struct {
+	// Namespace and Name of the Secret resource used to store the authority.
+	SecretNamespace, SecretName string
+
+	// RESTConfig used to connect to the apiserver.
+	RESTConfig *rest.Config
+
+	// The amount of time the root CA certificate will be valid for.
+	// This must be greater than LeafDuration.
+	// Defaults to 365d.
+	CADuration time.Duration
+
+	// The amount of time leaf certificates signed by ths authority will be
+	// valid for.
+	// This must be less than CADuration.
+	// Defaults to 7d.
+	LeafDuration time.Duration
+
+	// Logger to write messages to.
+	Log logr.Logger
+
+	lister corelisters.SecretNamespaceLister
+	client coreclientset.SecretInterface
+
+	// PEM-encoded CA certificate and private key bytes
+	currentCertData, currentPrivateKeyData []byte
+	// signMutex gates access to the certificate and private key data
+	signMutex sync.Mutex
+	// ensureMutex gates the 'ensureCA' method
+	ensureMutex sync.Mutex
+	// watchMutex gates access to the slice of watch channels
+	watchMutex sync.Mutex
+	watches    []chan struct{}
+}
+
+type SignFunc func(template *x509.Certificate) (*x509.Certificate, error)
+
+var _ SignFunc = (&DynamicAuthority{}).Sign
+
+func (d *DynamicAuthority) Run(stopCh <-chan struct{}) error {
+	if d.SecretNamespace == "" {
+		return fmt.Errorf("SecretNamespace must be set")
+	}
+	if d.SecretName == "" {
+		return fmt.Errorf("SecretName must be set")
+	}
+	if d.CADuration == 0 {
+		d.CADuration = time.Hour * 24 * 365 // 365d
```

Comment:
really Go... no time.Day :(

Reply:
A day doesn't have a consistent duration 😅 whereas hours (and below) do...